x-pack/legacy/plugins/epm/server/datasources/create.ts

```diff
@@ -29,31 +29,67 @@ export async function createDatasource(options: {
 }) {
   const { savedObjectsClient, callCluster, pkgkey, datasets, datasourceName, request } = options;
 
-  const packageInfo = await getPackageInfo({ savedObjectsClient, pkgkey });
-  if (packageInfo.status !== InstallationStatus.installed) {
+  const epmPackageInfo = await getPackageInfo({ savedObjectsClient, pkgkey });
+  if (epmPackageInfo.status !== InstallationStatus.installed) {
     throw new PackageNotInstalledError(pkgkey);
   }
-  const datasetNames = datasets.map(d => d.name);
-  const toSave = await installPipelines({ pkgkey, datasetNames, callCluster });
+
+  const registryPackageInfo = await Registry.fetchInfo(pkgkey);
+  // Pick the full dataset definition for each dataset name that has been requested
+  // from the package information from the registry.
+  // Requested dataset names that don't exist in the package will be silently ignored.
+  const datasetsRequestedNames = datasets.map(d => d.name);
+  const datasetsRequested = registryPackageInfo.datasets?.filter(packageDataset => {
+    return datasetsRequestedNames.includes(packageDataset.name);
+  });
+
+  const templateRefs: Array<Promise<AssetReference>> = [];
+  const pipelineRefs: Array<Promise<AssetReference[]>> = [];
+
+  if (datasetsRequested) {
+    datasetsRequested.forEach(dataset => {
+      const templateRef = installTemplateForDataset(
+        registryPackageInfo,
+        callCluster,
+        dataset,
+        datasourceName
+      );
+      if (templateRef) {
+        templateRefs.push(templateRef as Promise<AssetReference>); // Typescript thinks this may still be undefined here
+      }
+      if (dataset.ingest_pipeline) {
+        const pipelineRefArray = installPipelinesForDataset({
+          pkgkey,
+          dataset,
+          callCluster,
+          datasourceName,
+          packageName: registryPackageInfo.name,
```

Comment:
We hopefully can get rid of this line in the future with having `package` as part of the dataset info.